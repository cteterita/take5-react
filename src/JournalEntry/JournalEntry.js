import React from 'react';

function formatPrompts(prompts, complete) {
  return prompts.map((p) => {
    const { promptId, prompt, responses } = p;
    return (
      <fieldset
        id={promptId}
        key={promptId}
        disabled={complete}
      >
        {prompt}
        {responses.map((r, i) => {
          const id = `${i}-${p.promptId}`;
          return (
            <input
              key={id}
              id={id}
              defaultValue={r}
            />
          );
        })}
      </fieldset>
    );
  });
}

function JournalEntry(props) {
  const { type, saveEntry, entryData } = props;
  const handleSubmit = (e) => {
    e.preventDefault();
    const promptFields = e.target.querySelectorAll('fieldset');
    const newPrompts = [];
    promptFields.forEach((p) => {
      const responseFields = p.querySelectorAll('input');
      const responses = [];
      responseFields.forEach((r) => responses.push(r.value));
      newPrompts.push({
        prompt: p.innerText,
        promptId: Number(p.id),
        responses,
      });
    });
    saveEntry(type, newPrompts);
    e.target.reset();
  };

  if (!entryData) {
    return <span>Loading...</span>;
  }
  return (
    <form id={`${type}-form`} onSubmit={handleSubmit}>
      <h3>{type === 'morning' ? 'Morning Intentions' : 'Evening Reflections' }</h3>
      {formatPrompts(entryData.prompts, entryData.complete)}
      <button type="submit" disabled={entryData.complete}>Save</button>
    </form>
  );
}

export default JournalEntry;
