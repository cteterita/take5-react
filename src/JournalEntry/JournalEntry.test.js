import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';

import JournalEntry from './JournalEntry';

const exampleProps = {
  saveEntry: () => {},
  type: 'morning',
  entryData: {
    complete: true,
    prompts: [
      {
        prompt: 'Example prompt!',
        promptId: 1,
        responses: [
          'Example response 1',
          'Example response 2',
        ],
      },
    ],
  },
};

describe('JournalEntry component', () => {
  it('renders without errors', () => {
    const { saveEntry, type, entryData } = exampleProps;
    const div = document.createElement('div');
    ReactDOM.render(<JournalEntry saveEntry={saveEntry} type={type} entryData={entryData} />, div);
    ReactDOM.unmountComponentAtNode(div);
  });
  it('renders as expected', () => {
    const { saveEntry, type, entryData } = exampleProps;
    const tree = renderer.create(
      <JournalEntry saveEntry={saveEntry} type={type} entryData={entryData} />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
