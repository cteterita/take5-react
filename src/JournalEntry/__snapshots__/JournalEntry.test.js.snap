// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JournalEntry component renders as expected 1`] = `
<form
  id="morning-form"
  onSubmit={[Function]}
>
  <h3>
    Morning Intentions
  </h3>
  <fieldset
    disabled={true}
    id={1}
  >
    Example prompt!
    <input
      id="1-0"
      onChange={[Function]}
      value="Example response 1"
    />
    <input
      id="1-1"
      onChange={[Function]}
      value="Example response 2"
    />
  </fieldset>
  <fieldset>
    <button
      hidden={true}
      type="submit"
    >
      Save
    </button>
    <div
      className="morning user-feedback"
      hidden={true}
    />
  </fieldset>
</form>
`;
